Add fallback for ONVIF PTZ Stop with ContinuosMove and enhance PTZ operations

ramon-angosto opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
- Implement a fallback in STOP_MOVE to use ContinuousMove with zero velocity, resolving "Action Not Implemented" error during panning in some cameras. Solves error: ONVIF error when panning camera: "Failed to perform the action onvif/ptz. Action Not Implemented" #125680
-Introduce EMULATE_RELATIVE_MOVE to support cameras lacking RelativeMove using ContinuousMove with calculated sleep time. 
-Refine ABSOLUTE_MOVE to set speed values only for active axes, improving compatibility.


## Type of change

- [ ] Dependency upgrade
- [X ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #125680
- This PR is related to issue: 
- Link to documentation pull request: 
- Link to developer documentation pull request: 
- Link to frontend pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [X ] The code change is tested and works locally.
- [X ] Local tests pass. **Your PR cannot be merged unless tests pass**
- [X ] There is no commented out code in this PR.
- [X ] I have followed the [development checklist][dev-checklist]
- [X ] I have followed the [perfect PR recommendations][perfect-pr]
- [ ] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

ramon-angosto committed:
- Add fallback for ONVIF PTZ Stop and enhance PTZ operations (#125680)
- Implement a fallback in STOP_MOVE to use ContinuousMove with zero velocity, resolving "Action Not Implemented" error during panning.
-Introduce EMULATE_RELATIVE_MOVE to support cameras lacking RelativeMove using ContinuousMove with calculated sleep time.
-Refine ABSOLUTE_MOVE to set speed values only for active axes, improving compatibility.

ramon-angosto committed:
Merge branch 'dev' into fix-onvif

ramon-angosto committed:
Merge branch 'dev' into fix-onvif

jterrace commented on homeassistant/components/onvif/device.py:
```diff
@@ -542,12 +553,26 @@ async def async_perform_ptz(
+                    await ptz_service.Stop(
+                        {"ProfileToken": req.ProfileToken, "PanTilt": True, "Zoom": False}
+                    )
+                except:
```
what exception is expected here?

jterrace commented on homeassistant/components/onvif/device.py:
```diff
@@ -565,6 +590,46 @@ async def async_perform_ptz(
+                    req.Velocity = velocity
+                    await ptz_service.ContinuousMove(req)
+
+                except:
```
same here, can you use the exception you expect rather than a bare except?

jterrace commented on homeassistant/components/onvif/device.py:
```diff
@@ -577,11 +642,21 @@ async def async_perform_ptz(
+                if pan is not None:
+                    speed_pan_val = speed_val
+                if tilt is not None:
+                    speed_tilt_val = speed_val
```
you could simplify these a bit, e.g.

```
speed_pan_val = speed_val if pan else 0
```

jterrace commented on homeassistant/components/onvif/device.py:
```diff
@@ -607,8 +682,21 @@ async def async_perform_ptz(
-                await ptz_service.Stop(req)
+                try:
+                    await ptz_service.Stop(req)
+                except:
```
same question here about exceptions

ramon-angosto replied:
changed to:
                try:
                    await ptz_service.Stop(req)
                except Fault as e:
                    if "Action Not Implemented" in str(e):

ramon-angosto committed:
Addressed PR feedback:
- Added type annotations to calculate_sleep_time
- replaced 'in' with '==' using EMULATE_RELATIVE_MOVE constant
- specified Fault exception in async_perform_ptz
- simplified speed value assignment
- clarified exception handling for Stop command.

ramon-angosto committed:
Merge remote-tracking branch 'origin/fix-onvif' into fix-onvif

ramon-angosto committed:
Merge branch 'dev' into fix-onvif

ramon-angosto committed:
Addressed PR feedback:
- Added function is_action_not_implemented to handle Fault error "Action Not Implemented" error
- Correction of raise error in except to only raise if "Action Not Implemented" not in Fault error
Applied new helper function in `CONTINUOUS_MOVE`, `EMULATE_RELATIVE_MOVE`, and `STOP_MOVE` for consistent error handling.
- Fixed `STOP_MOVE` to only log continuous move usage when supported, ensuring accurate logging.
Addresses reviewer comments from PR #145811.

ramon-angosto committed:
Merge remote-tracking branch 'origin/fix-onvif' into fix-onvif

ramon-angosto committed:
Addressed PR feedback:
- Applied -> nit: flip the conditional to check "not in" and raise. That way, the rest of the error handling doesn't need to be indented
- Changed function from is_action_not_implemented to handle_is_action_not_implemented

ramon-angosto committed:
Merge branch 'dev' into fix-onvif

ramon-angosto reviewed (commented):
All comments made In the last pull request have been solved. New function has been implemented and the warning now only pops up if continous move is available

ramon-angosto committed:
Merge branch 'dev' into fix-onvif

ramon-angosto committed:
Merge branch 'dev' into fix-onvif

ramon-angosto committed:
Merge branch 'dev' into fix-onvif

ramon-angosto committed:
Merge branch 'dev' into fix-onvif

ramon-angosto reviewed (commented):
All changes have been reviewed and corrected

joostlek commented on homeassistant/components/onvif/device.py:
```diff
@@ -37,13 +37,28 @@
 from .models import PTZ, Capabilities, DeviceInfo, Profile, Resolution, Video
 
+def calculate_sleep_time(distance: float, full_rotation_time: float) -> float:
+    """
+    Calculate the sleep time based on the given distance and full rotation time.
+
+    Args:
+        distance (float): The distance to be covered.
+        full_rotation_time (float): The time required for a full rotation in seconds.
+
+    Returns:
+        float: The calculated sleep time in seconds.
+    """
```
we generally don't have this kind of docstrings

emontnemery replied:
We do allow it though, It's especially encouraged it when it's not obvious from the function name.
However, it's enough to have type information in the function protype, there's no need to duplicate it in the docstring.

joostlek commented on homeassistant/components/onvif/device.py:
```diff
@@ -37,13 +37,28 @@
+    Returns:
+        float: The calculated sleep time in seconds.
+    """
+    sleep_time_seconds = distance * full_rotation_time
+
+    return sleep_time_seconds
```
```suggestion
    return distance * full_rotation_time
```

joostlek commented on homeassistant/components/onvif/device.py:
```diff
@@ -522,9 +537,17 @@ async def async_perform_ptz(
             preset_val,
         )
+
+        def handle_is_action_not_implemented(error):
+            """Check if the error indicates that the action is not implemented."""
+            return "Action Not Implemented" not in str(error)
```
I would put this outside of the function. Also please add typing

joostlek commented on homeassistant/components/onvif/device.py:
```diff
@@ -565,6 +603,50 @@ async def async_perform_ptz(
+                    req = ptz_service.create_type(CONTINUOUS_MOVE)
+                    req.ProfileToken = profile.token
+                    if not profile.ptz or not profile.ptz.continuous:
+                        LOGGER.warning(
+                            "ContinuousMove not supported on device '%s'", self.name
+                        )
+                        return
```
So does this mean the service ends? I think instead we should raise a `ServiceValidationError` with this error. This way the user gets this as feedback instead of in the logs

joostlek reviewed (changes requested):
@jterrace Can you give this another pass? As I am by far not an ONVIF expert, I would like a second opinion from you here
